### PR TITLE
use fast base58 encode implementation (20~60% performance gain)

### PR DESCRIPTION
**Pull request by rubensayshi**
```
$ go test -bench=. base58/base58bench_test.go 
BenchmarkBase58FastEncode-7        	      30	  45674986 ns/op	   0.11 MB/s
BenchmarkBase58SlowEncode-7        	      20	  63573231 ns/op	   0.08 MB/s
BenchmarkBase58FastEncodeSmall-7   	 1000000	      1712 ns/op	  17.52 MB/s
BenchmarkBase58SlowEncodeSmall-7   	  300000	      5336 ns/op	   5.62 MB/s
BenchmarkBase58Decode-7            	     300	   5276054 ns/op	   1.29 MB/s
```

### CODE REVIEW

**Comment by dajohi** [comment]
Hi @rubensayshi 

Can you please rebase and fix the Travis CI errors?

**Comment by jrick** [comment]
And while it's useful for the benchmarks, now that we've seen them, there's no reason to keep the slow implementation around.

**Comment by stevenroose** [comment]
Indeed. A `FastEncode` method is really not necessary. Just replace the existing encode implementation with the fast one.

**Comment by davecgh** [comment]
Thanks for the PR.  As others have mentioned, I'd really prefer just replacing the `Encode` with the new faster method.  If you would please do that and rebase/squash everything down, I'll get this approved and merged.

**Comment by rubensayshi** [comment]
@davecgh okay, glad to see it's well received, I'll cleanup the PR ;)

File: base58/base58bench_test.go [hunk 1]
```diff
@@ -11,15 +11,21 @@ import (
 	"github.com/btcsuite/btcutil/base58"
 )
 
-func BenchmarkBase58Encode(b *testing.B) {
+var result string
+var resultb []byte
+
+func BenchmarkBase58FastEncode(b *testing.B) {
```
Comment (jrick) [inline base58/base58bench_test.go]: `BenchmarkBase58Encode`

File: base58/base58bench_test.go [hunk 2]
```diff
@@ -11,15 +11,21 @@ import (
...
 	b.StopTimer()
 	data := bytes.Repeat([]byte{0xff}, 5000)
 	b.SetBytes(int64(len(data)))
 	b.StartTimer()
 
+	var s string
 	for i := 0; i < b.N; i++ {
-		base58.Encode(data)
+		s = base58.Encode(data)
```
Comment (davecgh) [inline base58/base58bench_test.go]: Please revert this change.  It is not necessary and does not change the results at all.

With:
```
$ go test -bench=BenchmarkBase58 -benchmem
BenchmarkBase58FastEncode-6           20          71934135 ns/op           0.07 MB/s       20736 B/op          3 allocs/op
```

Without:
```
$ go test -bench=BenchmarkBase58 -benchmem
BenchmarkBase58FastEncode-6           20          72284180 ns/op           0.07 MB/s       20736 B/op          3 allocs/op
```

File: base58/base58bench_test.go [hunk 3]
```diff
@@ -29,7 +35,10 @@ func BenchmarkBase58Decode(b *testing.B) {
 	b.SetBytes(int64(len(encoded)))
 	b.StartTimer()
 
+	var bs []byte
 	for i := 0; i < b.N; i++ {
-		base58.Decode(encoded)
+		bs = base58.Decode(encoded)
```
Comment (davecgh) [inline base58/base58bench_test.go]: Same here.  Please revert unnecessary change.

**Comment by rubensayshi** [comment]
whoops, almost forgot the important stuff!  
the implementation is from https://github.com/mr-tron/go-base58 MIT license 

how to properly incorperate this?

**Comment by jrick** [comment]
For the license, you can simply copy it into the source file.

**Comment by davecgh** [comment]
Oh, I didn't realize it had another license.  Everything in btcsuite must be licensed under the ISC license.  Unfortunately, I'm going to have to decline it as is unless you can get permission to relicense it.

**Comment by stevenroose** [comment]
@davecgh Isn't the ISC license totally equivalent to the MIT license?

> The ISC license is a permissive free software license written by the Internet Systems Consortium (ISC). It is meant to be functionally equivalent to the simplified BSD and the MIT licenses, differing in its removal of language deemed unnecessary following the global adoption of the Berne Convention.

**Comment by mr-tron** [comment]
I gave permission special for this project.

**Comment by rubensayshi** [comment]
@davecgh @mr-tron here is the original author, is that enough or you need some more explicit?

**Comment by davecgh** [comment]
Great.  Thanks @mr-tron.

@rubensayshi Thanks for following up.  The author has provided permission.  That's all we needed.

File: base58/base58.go [hunk 4]
```diff
@@ -46,30 +48,43 @@ func Decode(b string) []byte {
...
+
+	for zcount < binsz && bin[zcount] == 0 {
+		zcount++
 	}
 
-	// leading zero bytes
-	for _, i := range b {
-		if i != 0 {
-			break
+	size := (binsz-zcount)*138/100 + 1
```
Comment (rubensayshi) [inline base58/base58.go]: I wish I knew, but I don't .... 

the implementation is based off this C implementation: https://github.com/trezor/trezor-crypto/blob/master/base58.c
but that has no explanation for it either.

Comment (mr-tron) [inline base58/base58.go]: It's probably misprint from original algorithm.  137 is enough.
https://searchcode.com/codesearch/view/94506854/
   // Expected size increase from base58 conversion is approximately 137%
    // use 138% to be safe

[the remark this answers is not in the document]
Comment (stevenroose) [reply base58/base58.go]: @rubensayshi Last nit to get this merged :)

**Comment by ribasushi** [comment]
For those following along: I opened a new PR based on a newer iteration of pretty much the same code. Enc/Dec of b58 addresses is 4+/7+ **times** faster: https://github.com/btcsuite/btcutil/pull/170#issue-426748886

**Comment by rubensayshi** [comment]
👍  @ribasushi , I'll close my PR and trust you'll pull yours to the finish line :)

**Comment by ribasushi** [comment]
@rubensayshi I am sorry your PR got stuck, not sure why it happened :(
If time permits and you happen to have a more extensive test suite: it would be great if you test the proposed changes a bit more. You know... just to be sure ;)

**Comment by rubensayshi** [comment]
no worries, it happens and I didn't bother reviving it so it's on me ;)

I don't have any more tests and no longer work with the codebase that was using this either so can't run it through that testsuite sorry
